File: frontend/src/api/summary.ts
import { handleUnauthorized } from './auth'

export type SummaryDays = '7' | '30'

export type SummaryPoint = { date: string; value: number }
export type SummaryBpPoint = { date: string; systolic: number; diastolic: number }
export type SummaryWalkPoint = { date: string; distance: number; time_seconds: number | null; stops: number | null }
export type SummarySymptomsEntry = { date: string; values: string[] }
export type SummaryNoteEntry = { date: string; text: string }

export type SummaryResponse = {
  range: { days: number; start_date: string; end_date: string; generated_at: string }
  vitals: { weight: SummaryPoint[]; pulse: SummaryPoint[]; bp: SummaryBpPoint[] }
  activity: { walk: SummaryWalkPoint[]; songs: SummaryPoint[] }
  functional: { nyha: SummaryPoint[] }
  symptoms: SummarySymptomsEntry[]
  notes: SummaryNoteEntry[]
}

async function parseJson<T>(response: Response): Promise<T> {
  if (!response.ok) {
    if (response.status === 401) handleUnauthorized()
    throw new Error(String(response.status))
  }
  return response.json() as Promise<T>
}

export async function getSummary(
  accessToken: string,
  days: SummaryDays,
): Promise<SummaryResponse> {
  const response = await fetch(`/api/summary?days=${days}`, {
    headers: { Authorization: `Bearer ${accessToken}` },
    credentials: 'include',
  })
  return parseJson<SummaryResponse>(response)
}

File: backend/app/schemas/summary.py
from pydantic import BaseModel


class DumpableModel(BaseModel):
    def __eq__(self, other: object) -> bool:
        if isinstance(other, dict):
            return self.model_dump() == other
        return super().__eq__(other)


class SummaryPoint(DumpableModel):
    date: str
    value: float | int


class SummaryBpPoint(DumpableModel):
    date: str
    systolic: int
    diastolic: int


class SummaryWalkPoint(DumpableModel):
    date: str
    distance: int
    time_seconds: int | None = None
    stops: int | None = None


class SummarySymptomsEntry(DumpableModel):
    date: str
    values: list[str]


class SummaryNoteEntry(DumpableModel):
    date: str
    text: str


class SummaryRange(BaseModel):
    days: int
    start_date: str
    end_date: str
    generated_at: str


class SummaryVitals(BaseModel):
    weight: list[SummaryPoint]
    pulse: list[SummaryPoint]
    bp: list[SummaryBpPoint]


class SummaryActivity(BaseModel):
    walk: list[SummaryWalkPoint]
    songs: list[SummaryPoint]


class SummaryFunctional(BaseModel):
    nyha: list[SummaryPoint]


class SummaryResponse(BaseModel):
    range: SummaryRange
    vitals: SummaryVitals
    activity: SummaryActivity
    functional: SummaryFunctional
    symptoms: list[SummarySymptomsEntry]
    notes: list[SummaryNoteEntry]
